### DISCUSSION TITLE
Custom Actions

### DISCUSSION THREAD

**Discussion by VinayaSathyanarayana** [keystonejs/keystone#5315] [Feature requests]
There was a discussion on Custom Actions  in the earlier versions of Keystone. 

We need to enable Custom Actions on Lists (Both List View and Item View)
``` javascript
keystone.createList('ListKey', {
  fields: { ... },
  ui: { ... },
  customActions:  [
    {  name: "Custom Action 1", displayMode: "Button",  access: () => {...}, action: (args) => {...} },
    {  name: "Custom Action 2", displayMode: "Button",  access: () => {...}, action: (args) => {...} },
  ]
});    
```

The action functions should take context and the list of selected ids of the List Items on which the custom action has to be performed.

The Custom Action should be available in both List and Item Views based on access privileges
In the List View, the custom actions will get a list of selected ids
In the Item View, the custom actions will get only one id

@JedWatson - Your Views please

**Comment by VinayaSathyanarayana** [comment]
Any views on this @JedWatson @timleslie

**Comment by ChuckJonas** [comment]
I second the need for this... 

Really the entire layout should be configurable:

- Configuration/Data driven "layouting" of the fields
- Ability to add include react custom react components

Something to this effect:
```
(item, context) => {
   header: // allow custom react component
   actions: ...
   footer: ....
   fieldForm: [
      // all both keystone fields and custom react components
   ]
}
```

There should also be support for multiple layouts or some way to switch the layout based on the page context & item

**Comment by raveling** [comment]
Thanks @VinayaSathyanarayana for raising this. 🙇  We discussed **custom actions** recently, and have it noted as one of many improvements we can bring to Admin UI.

@ChuckJonas your suggestions to **make the entire layout configurable** , and **have support for multiple layouts** have also been talked about, and are (generally speaking) aligned with what we're thinking for the future. 🙌 

Both of your ideas effect, and are effected by, other improvements we have planned for Admin UI. The key here is to solve them holistically so that the Admin UI customisation experience is really intuitive no matter what you want to achieve.

Definitely something we can do, but needs proper thought to get it right at the system level. This kind of work won't begin in earnest until after we complete the `General Availability` release for Keystone.

Thanks!

**Reply by rburgst** [reply]
is there a timeline or a priorization so we know what to expect?

**Comment by VinayaSathyanarayana** [comment]
Any updates on this?

**Comment by jerguslejko** [comment]
here goes another: any updates on this?

**Reply by VinayaSathyanarayana** [reply]
Please share any updates

**Comment by dcousens** [comment]
Please see https://github.com/keystonejs/keystone/pull/9689 💙
